packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -690,8 +804,22 @@ class TextPainter {
   ///
   /// Each box corresponds to a [PlaceholderSpan] in the order they were defined
   /// in the [InlineSpan] tree.
-  List<TextBox>? get inlinePlaceholderBoxes => _inlinePlaceholderBoxes;
-  List<TextBox>? _inlinePlaceholderBoxes;
+  List<TextBox>? get inlinePlaceholderBoxes {
+    final _TextPainterLayoutCache? layout = _layoutCache;
+    if (layout == null) {
+      return null;
+    }
+    final List<TextBox> rawBoxes = layout._inlinePlaceholderBoxes
+                               ??= layout.paragraph.getBoxesForPlaceholders();
```

Comment:
this might be clearer?
```suggestion
    layout._inlinePlaceholderBoxes ??= layout.paragraph.getBoxesForPlaceholders();
    final List<TextBox> rawBoxes = layout._inlinePlaceholderBoxes;
```

Reply:
`_inlinePlaceholderBoxes`'s type is `List<TextBox>?` so I would have to use `!`.